tools/shared/src/main/scala/org/scalajs/core/tools/io/IRFileCache.scala

```diff
@@ -0,0 +1,314 @@
+      val result = Seq.newBuilder[PersistedFiles]
+
+      for (file <- files) {
+        @tailrec
+        def putContents(): PersistedFiles = {
+          val newValue = new PersistedFiles(file.path)
+          val oldValue = globalCache.putIfAbsent(file.path, newValue)
+
+          val contents = if (oldValue != null) oldValue else newValue
+
+          if (contents.reference()) contents
+          else putContents()
+        }
+
+        val contents = putContents()
+        contents.update(file)
+        result += contents
+      }
+
+      free()
+      localCache = result.result()
+    }
+
+    /** Should be called if this cache is not used anymore.
+     *
+     *  Frees resources in the global cache, if they are not used anymore.
+     *  The cache may be reused after calling [[free]] (but this is not any
+     *  faster than calling [[newCache]], modulo the object allocation).
+     */
+    def free(): Unit = {
+      if (localCache != null) {
+        localCache.foreach(_.dereference())
+        localCache = null
+      }
+    }
+
+    protected override def finalize(): Unit = {
+      free()
+      super.finalize()
+    }
+  }
+
+  /** Stores the extracted [[VirtualScalaJSIRFile]]s from the file at path.
+   *
+   *  This also tracks references to itself by reference counting.
+   *  Further, a [[PersistedFiles]] has a tombstone state. It is necessary to
+   *  avoid a race between referencing a file just retrieved from
+   *  [[globalCache]] and removing a file from [[globalCache]] that has just
+   *  been dereferenced.
+   */
+  private final class PersistedFiles(path: String) {
+
+    /** Number of references we have. -1 means we are a tombstone */
+    private[this] val _references = new AtomicInteger(0)
+
+    /** Last version we have been updated with.
+     *  May only be written under synchronization, except this is a tombstone
+     */
+    @volatile
+    private[this] var _version: Option[String] = None
+
+    /** Files in this [[PersistedFiles]]
+     *  May only be written under synchronization, except this is a tombstone
+     */
+    @volatile
+    private[this] var _files: Seq[VirtualRelativeIRFile] = null
+
+    def files: Seq[VirtualRelativeIRFile] = _files
+
+    /** Try to reference this block of files.
+     *  @return true if referencing succeeded, false if this is a tombstone
+     */
+    @tailrec
+    final def reference(): Boolean = {
+      val refs = _references.get
+
+      if (refs == -1) {
+        // we are a tombstone, help cleaning up and bail out
+        cleanup()
+        false
+      } else {
+        // try to increase ref count
+        if (_references.compareAndSet(refs, refs + 1)) true // done
+        else reference() // something changed, try again
+      }
+    }
+
+    /** Dereference this file.
+     *
+     *  If there are no references any more, turn this [[PersistedFiles]] into
+     *  a tombstone and remove it from the cache.
+     */
+    final def dereference(): Unit = {
```

Comment:
The name is weird, as it looks like "dereference a pointer". I suggest `unreference`.